### PR TITLE
[fix] OSS restore state to proper device

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes the state pull and push behavior to being something more easily predictable, when a state is pulled it gets moved to CPU (failsafe if this is a reasonably big model). Upon restore the `param_groups` were not restored to the proper device. It looks like there are still probable duplicates in that field though

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
         self.load_local_state_dict(state_dict["state"][self.rank])
 
         # Restore the global param_groups
-        self.param_groups = state_dict["param_groups"]
+        self.param_groups = recursive_copy_to_device(state_dict["param_groups"], non_blocking=True, device=self._device)
```
Comment (min-xu-ai) [inline fairscale/optim/oss.py:151]: Line too long?

Is there a test for this already? If so, should there be some assert in the test?

Comment (blefaudeux) [reply fairscale/optim/oss.py:151]: no test, it's a good point, I'll add that

Comment (msbaines) [reply fairscale/optim/oss.py:151]: We follow PyTorch coding style of 120 columns:

https://github.com/pytorch/pytorch/blob/master/.flake8#L3

It is enforced here:

https://github.com/facebookresearch/fairscale/blob/master/setup.cfg#L29

Comment (blefaudeux) [reply fairscale/optim/oss.py:151]: it's supposed to be automatic with vscode+black, it's pure FB tooling and 120 lines, this formatting crap is so annoying.. I'll fix that in the next PR

Comment (blefaudeux) [reply fairscale/optim/oss.py:151]: ah no ok, black is happy with that indeed, it's 121 cols but for some reason this passes (same for the init above, also 121). Anyway, all good then

**Pull request merged**